lib/view_component/test_helpers.rb

```diff
@@ -21,13 +21,27 @@ def refute_component_rendered
       if ENV["DEBUG"]
         warn(
           "WARNING in `ViewComponent::TestHelpers`: Add `capybara` " \
-          "to Gemfile to use Capybara assertions."
+            "to Gemfile to use Capybara assertions."
         )
       end
 
       # :nocov:
     end
 
+    def self.included(base)
+      if defined?(Devise::Test::ControllerHelpers)
+        # the `@request` instance variable is used by Devise::Test::ControllerHelpers and must
+        # be in a `setup` block before including `Devise::Test::ControllerHelpers`
+        base.instance_eval do
+          setup do
+            @request = __vc_render_preview_controller.request
+          end
+        end
+
+        base.include(Devise::Test::ControllerHelpers)
```

Comment:
I don't think we want to include third party modules into the ViewComponent test helper by default.

Can we detect if devise has already included these helpers or not?

Reply:
probably

Reply:
before I look for an alternative, I'd contend we do exactly that on line 8. Doing it this way would certainly make things less difficult on the end user.

Reply:
I'm not certain we need to include this manually at all. I looked at the devise source a bit and I think we could get away with something similar to what you have here, but maybe modified a bit. Thoughts?

```ruby
def before_setup
  super
  _vc_reset!
end

def _vc_reset!
  @request = ActionDispatch::TestRequest.create
  @request.session = ActionController::TestSession.new
  @request
end
```

Reply:
As it turns out, Devise and Warden hook rather deeply into the controller, and `ActionDispatch::TestRequest` wont provide the necessary context. I simplified my implementation to:

```ruby
 def before_setup
      @request = __vc_render_preview_controller.request
      super
 end
 ```

The request must be set up before `super` is called, otherwise a hook in the devise test helpers will fail.